### ISSUE TITLE
Adding support and/or utilities for parallel reads over BatchingVisitibles iterables

### ISSUE THREAD

**Issue by jnelson15** [palantir/atlasdb#1706]
When making an initial `getRanges` call on Cassandra, Atlas parallelizes fetching the first pages of the `BatchingVisitable`s up to the `rangesConcurrency` value. After the initial page, it seems like the burden of parallelization is up to the user. However, most users seem to use the `BatchingVisitibles#concat` or simply just loop over the iterable to get results, therefore serializing the calls. It may be worth adding some utilities to help do this. If we could do it using the same executor that the `CassandraKeyValueService#getFirstBatchForRanges` call uses that would be ideal since it's already configured for a similar use-case. However, since this is Cassandra-specific I imagine this could be hard, so alternatively we could just add some parallel versions of the `BatchingVisitibles` classes that take in executors from the client. The former is a little safer, though, since the `CassandraKeyValueService` pool ensures that it is the same size as the number of underlying connections so we can't accidentally exhaust our pool as easily.

**Comment by jnelson15** [comment]
@tpetracca @jboreiko am I understanding these correctly? Main concern here is the assumption by most users that their large `getRanges` calls are parallel when they're actually not.

**Comment by rhero** [comment]
Going to backlog this issue for now since it's unlikely we're going to take a look at doing this within the next few weeks, but we're about to do some longer term planning and may include taking a look at the BatchingVisitibles API.

**Comment by tpetracca** [comment]
Given how confusing it is, I'm want to just throw an `@Deprecated` on `getRanges` given that it's basically impossible to use the way people want to use it (from what I can tell), and only adds confusion to understanding how a thing is actually executing. In practice I plan to remove it as I see it from internal products.

**Comment by jnelson15** [comment]
What would be the non-deprecated version? Letting the user control their own parallelism and doing individual `getRange` calls?

**Comment by jnelson15** [comment]
If we can implement https://github.com/palantir/atlasdb/issues/1738 well then the `getRanges` seems sensible, but we would also want to complement it with a better way to get the rest of the batches in parallel as well.

**Comment by tpetracca** [comment]
Yea, @jnelson15. I think the better approach is to use known serial or parallel invocation via things like `Iterables.transform` or `Iterator.forEach` or your good old fashioned for loop.

**Comment by jnelson15** [comment]
Yeah that makes sense. May be worthwhile adding some utilities to help do that to avoid lots of code duplication but that'll be a lot more clear than the existing situation where it's kind of blurred between the different abstraction layers (KVS code vs Batching Visitibles utilities).

**Comment by tpetracca** [comment]
Yea I think seeing how people opt in to using more explicit methods of serial vs. parallel will give us a better sense of desired usage for what we do decide to revamp this stuff and build a better API.

**Comment by jnelson15** [comment]
@jboreiko and I just talked a bit more about this and https://github.com/palantir/atlasdb/issues/1738. Bunch of our microservices products have been hit by missing the fact that `getRanges` pre-fetches the first page of each range. Sounds like we should pull the pre-fetching out (other than maybe for a large internal product that depends on the concurrency) and then make the parallelism more user-facing as described above. Closing out the other ticket since they're kind of converging a bit.
